### ISSUE TITLE
PHP Fatal error:  Uncaught Error: Call to a member function emit() on null

### ISSUE THREAD

**Issue by crabovwik** [friends-of-reactphp/mysql#70]
Hello.

I am getting a fatal error on version 0.3.3.

But all works fine on 0.3.2. 

Something gone wrong in Parse. I attached a log trace below.

```
PHP Fatal error:  Uncaught Error: Call to a member function emit() on null in /var/www/huj/public_html/libraries/vendor/react/mysql/src/Protocal/Parser.php:322
#0 /var/www/huj/public_html/libraries/vendor/react/mysql/src/Protocal/Parser.php(309): React\MySQL\Protocal\Parser->onResultRow(Array)
#1 /var/www/huj/public_html/libraries/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(123): React\MySQL\Protocal\Parser->parse('\x01\x00\x00\x01'R\x00\x00\x02\x03def\fc...')
#2 /var/www/huj/public_html/libraries/vendor/react/stream/src/Util.php(71): Evenement\EventEmitter->emit('data', Array)
#3 /var/www/huj/public_html/libraries/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(123): React\Stream\Util::React\Stream\{closure}('\x01\x00\x00\x01'R\x00\x00\x02\x03def\fc...')
#4 /var/www/huj/public_html/libraries/vendor/react/stream/src/DuplexResourceStream.php(193): Evenement\EventEmitter->emit('data', Array)
#5 [internal function]: React\Stream\DuplexResourceStream->handleData(Resource id #397)
#6 / in /var/www/huj/public_html/libraries/vendor/react/mysql/src/Protocal/Parser.php on line 322

Fatal error: Uncaught Error: Call to a member function emit() on null in /var/www/huj/public_html/libraries/vendor/react/mysql/src/Protocal/Parser.php on line 322

Error: Call to a member function emit() on null in /var/www/huj/public_html/libraries/vendor/react/mysql/src/Protocal/Parser.php on line 322

Call Stack:
    0.0004     398984   1. {main}() /var/www/huj/public_html/libraries/WebSocketServer/web-socket-server.php:0
    0.1307   13110064   2. Ratchet\Server\IoServer->run() /var/www/huj/public_html/libraries/WebSocketServer/web-socket-server.php:156
    0.1307   13110064   3. React\EventLoop\StreamSelectLoop->run() /var/www/huj/public_html/libraries/vendor/cboden/ratchet/src/Ratchet/Server/IoServer.php:74
  940.7393   16629272   4. React\EventLoop\StreamSelectLoop->waitForStreamActivity() /var/www/huj/public_html/libraries/vendor/react/event-loop/src/StreamSelectLoop.php:205
  940.7393   16632312   5. call_user_func:{/var/www/huj/public_html/libraries/vendor/react/event-loop/src/StreamSelectLoop.php:238}() /var/www/huj/public_html/libraries/vendor/react/event-loop/src/StreamSelectLoop.php:238
  940.7393   16632312   6. React\Stream\DuplexResourceStream->handleData() /var/www/huj/public_html/libraries/vendor/react/event-loop/src/StreamSelectLoop.php:238
  940.7393   16702344   7. React\Stream\DuplexResourceStream->emit() /var/www/huj/public_html/libraries/vendor/react/stream/src/DuplexResourceStream.php:193
  940.7393   16702344   8. React\Stream\Util::React\Stream\{closure}() /var/www/huj/public_html/libraries/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php:123
  940.7393   16702720   9. React\Socket\Connection->emit() /var/www/huj/public_html/libraries/vendor/react/stream/src/Util.php:71
  940.7393   16702720  10. React\MySQL\Protocal\Parser->parse() /var/www/huj/public_html/libraries/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php:123
  940.7432   16745824  11. React\MySQL\Protocal\Parser->onResultRow() /var/www/huj/public_html/libraries/vendor/react/mysql/src/Protocal/Parser.php:309
```

**Comment by clue** [comment]
@crabovwik Thank you for reporting!

Unfortunately, I'm currently unable to reproduce the problem you're seeing.

Is the problem you're seeing reproducible? Does this only happen for certain queries, perhaps certain data types or column lengths? Can you send me (an extract of) the data that is causing these issues for you?

**Comment by crabovwik** [comment]
Thanks for the answer, @clue!

Yes, it's reproducible.

I'll give additional information about used queries and mysql tables. Today / tomorrow.

**Comment by valVk** [comment]
Personally I can get exact the same issue when use pool connection and amount of call more than available connections. BTW code which I'm using is in PR  https://github.com/friends-of-reactphp/mysql/pull/27

**Comment by clue** [comment]
Unfortunately, I'm currently unable to reproduce the problem you're seeing.

Is the problem you're seeing reproducible? Does this only happen for certain queries, perhaps certain data types or column lengths? Can you send me (an extract of) the data that is causing these issues for you?

Can you try running this against current master branch? Does this only affect the last tagged release?

**Comment by adriaangraas** [comment]
Hi, this problem occurs to me as well in some complicated cases of nesting of promises. I'm on the master branch.

```php
$cp = $mysqlFactory->createConnection("...");

$cp->then(function (ConnectionInterface $c) {
    return $c->ping()->then(function () use ($c) {
        return $c;
    });
})->then(function() use ($cp) {
    $cp->then(function ($c) {
        $c->query("select * from whatever");
    });
});

$cp->then(function() use ($cp) {
    $cp->then(function ($c) {
        $c->query("select * from whatever");
    });
})->then(function() {
   var_dump("Never arrives here.");
});
```

**Comment by adriaangraas** [comment]
Hi, I am encountering the same problem. The problem seems to occur when there are nested promises with asynchronous access to the connection.

```php
        $cp->then(function (ConnectionInterface $c) {
            return $c->ping();
        })->then(function () use ($cp) {
            $cp->then(function (ConnectionInterface $c) {
                $c->ping();
            });
        });

        $cp->then(function (ConnectionInterface $c) {
            return $c->ping();
        })->then(function () {
            var_dump("Never arrives here.");
            exit;
        });
```

**Comment by clue** [comment]
@adriaangraas Thank you for providing this gist! We can now reproduce the issue you're seeing and I'm working on a fix for this :+1:

**Comment by elprawn** [comment]
I've created a PR for the issue https://github.com/friends-of-reactphp/mysql/pull/74 with an explanation.

Essentially, we dequeue a command and make it the current one prematurely, because we null the current command before calling the promise's resolver (even though that may also contain a query).

**Comment by clue** [comment]
@elprawn Thank you for looking into this! I've also just filed #75 which also addresses the very same issue in a very similar way and also includes a test to ensure this will never pop up again :+1: 

This was caused by the `Parser` trying to execute multiple commands at the same time when commands are already queued and a new command is queued from a success handler (thank you @adriaangraas for helping figure this out!). When the second command was completed, it would no longer have a reference to the command and would bail out. This is a side effect that was introduced with #52, but the underlying flaw has been present for much longer. The fix ensures we only ever take one command from the queue and then wait for it to finish.
